Library/Homebrew/keg.rb

```diff
@@ -101,6 +102,19 @@ def initialize(path)
     @name = path.parent.basename.to_s
     @linked_keg_record = HOMEBREW_LIBRARY.join("LinkedKegs", name)
     @opt_record = HOMEBREW_PREFIX.join("opt", name)
+    begin
+      tab = Tab.from_file(path/"INSTALL_RECEIPT.json")
+      tap = tab.tap
+
+      oldname = if tap.nil? || tap == "Homebrew/homebrew"
+        Formulary.factory(name)
+      else
+        Formulary.factory("#{tap.sub("homebrew-", "")}/#{name}")
+      end.oldname
+
+      @oldname_opt_record = HOMEBREW_PREFIX.join("opt", oldname) if oldname
+    rescue Exception
+    end
```

Comment:
Yes, I see your idea, but this still doesn't work. We need to optlink oldname, when it's not linked (e.g. on `upgrade`. First of all we unlink linked keg in `upgrade`, then we install new keg, and then link it. If we try to optlink that new keg, there won't be any symlink in `opt` pointing to the new keg, so we have to use `path.parent == dir.resolved_path.parent` to know that it point to the right package).